json/value.go

```diff
@@ -0,0 +1,172 @@
+package json
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"math"
+	"math/big"
+	"strconv"
+)
+
+// Value represents a JSON Value type
+// JSON Value types: Object, Array, String, Number, Boolean, and Null
+type Value struct {
+	w       *bytes.Buffer
+	scratch *[]byte
+}
+
+// newValue returns a new Value encoder
+func newValue(w *bytes.Buffer, scratch *[]byte) Value {
+	return Value{w: w, scratch: scratch}
+}
+
+// String encodes v as a JSON string
+func (jv Value) String(v string) {
+	escapeStringBytes(jv.w, []byte(v))
+}
+
+// Byte encodes v as a JSON number
+func (jv Value) Byte(v int8) {
+	jv.Long(int64(v))
+}
+
+// Short encodes v as a JSON number
+func (jv Value) Short(v int16) {
+	jv.Long(int64(v))
+}
+
+// Integer encodes v as a JSON number
+func (jv Value) Integer(v int32) {
+	jv.Long(int64(v))
+}
+
+// Long encodes v as a JSON number
+func (jv Value) Long(v int64) {
+	*jv.scratch = strconv.AppendInt((*jv.scratch)[:0], v, 10)
+	jv.w.Write(*jv.scratch)
+}
+
+// Float encodes v as a JSON number
+func (jv Value) Float(v float32) {
+	jv.float(float64(v), 32)
+}
+
+// Double encodes v as a JSON number
+func (jv Value) Double(v float64) {
+	jv.float(v, 64)
+}
+
+func (jv Value) float(v float64, bits int) {
+	*jv.scratch = encodeFloat((*jv.scratch)[:0], v, bits)
+	jv.w.Write(*jv.scratch)
+}
+
+// Boolean encodes v as a JSON boolean
+func (jv Value) Boolean(v bool) {
+	*jv.scratch = strconv.AppendBool((*jv.scratch)[:0], v)
+	jv.w.Write(*jv.scratch)
+}
+
+// EncodedBytes writes v as a base64 value in JSON string
+func (jv Value) EncodedBytes(v []byte) {
+	encodeByteSlice(jv.w, (*jv.scratch)[:0], v)
+}
+
+// WriteBytes writes v directly to the JSON document
+func (jv Value) WriteBytes(v []byte) {
+	jv.w.Write(v)
+}
+
+// Array returns a new Array encoder
+func (jv Value) Array() *Array {
+	return newArray(jv.w, jv.scratch)
+}
+
+// Object returns a new Object encoder
+func (jv Value) Object() *Object {
+	return newObject(jv.w, jv.scratch)
+}
+
+// Null encodes a null JSON value
+func (jv Value) Null() {
+	jv.w.WriteString(null)
+}
+
+// BigInteger encodes v as JSON value
+func (jv Value) BigInteger(v *big.Int) {
+	jv.w.Write([]byte(v.Text(10)))
+}
+
+// BigDecimal encodes v as JSON value
+func (jv Value) BigDecimal(v *big.Float) {
+	if i, accuracy := v.Int64(); accuracy == big.Exact {
+		jv.Long(i)
+		return
+	}
+	jv.w.Write([]byte(v.Text('e', -1)))
```

Comment:
Curious why the `'e', -1` is needed?

Reply:
By default `String()` will output using `Text('g', 10)` which limits precision to 10 digits, and will may output a value in decimal or exponent depending on the size.